core/src/main/java/org/bitcoinj/base/VarInt.java

```diff
@@ -31,17 +33,54 @@ public class VarInt {
      *
      * @param value the unsigned long value (beware widening conversion of negatives!)
      */
-    public VarInt(long value) {
-        this.value = value;
-        originallyEncodedSize = getSizeInBytes();
+    public static VarInt of(long value) {
+        return new VarInt(value, sizeOf(value));
     }
 
     /**
      * Constructs a new VarInt with the value parsed from the specified offset of the given buffer.
      *
      * @param buf the buffer containing the value
      * @param offset the offset of the value
+     * @throws ArrayIndexOutOfBoundsException if offset points outside of the buffer, or
+     *                                        if the value doesn't fit the remaining buffer
      */
+    public static VarInt ofBytes(byte[] buf, int offset) throws ArrayIndexOutOfBoundsException {
+        check(offset >= 0 && offset < buf.length, () ->
+                new ArrayIndexOutOfBoundsException(offset));
+        int first = 0xFF & buf[offset];
+        long value;
+        int originallyEncodedSize;
+        if (first < 253) {
+            value = first;
+            originallyEncodedSize = 1; // 1 data byte (8 bits)
+        } else if (first == 253) {
+            value = ByteUtils.readUint16(buf, offset + 1);
+            originallyEncodedSize = 3; // 1 marker + 2 data bytes (16 bits)
+        } else if (first == 254) {
+            value = ByteUtils.readUint32(buf, offset + 1);
+            originallyEncodedSize = 5; // 1 marker + 4 data bytes (32 bits)
+        } else {
+            value = ByteUtils.readInt64(buf, offset + 1);
+            originallyEncodedSize = 9; // 1 marker + 8 data bytes (64 bits)
+        }
+        return new VarInt(value, originallyEncodedSize);
+    }
```

Comment:
I was expecting to see this code removed elsewhere, but I understand the reasons you didn't do it.

Reply:
Because the now deprecated native constructor cannot call the new static constructor.

We could craft a third static helper method, but I figured this isn't worth it.